pkg/controller/build/build_controller.go

```diff
@@ -791,191 +826,211 @@ func (ctrl *Controller) postBuildCleanup(pool *mcfgv1.MachineConfigPool, ignoreM
 	)
 }
 
-// Marks a given MachineConfigPool as build successful and cleans up after itself.
-func (ctrl *Controller) markBuildSucceeded(ps *poolState) error {
-	klog.Infof("Build succeeded for MachineConfigPool %s, config %s", ps.Name(), ps.CurrentMachineConfig())
-
-	pool := ps.MachineConfigPool()
-
-	// Get the final image pullspec.
-	imagePullspec, err := ctrl.imageBuilder.FinalPullspec(pool)
-	if err != nil {
-		return fmt.Errorf("could not get final image pullspec for pool %s: %w", ps.Name(), err)
-	}
-
-	if imagePullspec == "" {
-		return fmt.Errorf("image pullspec empty for pool %s", ps.Name())
+// If one wants to opt out, this removes all of the statuses and object
+// references from a given MachineConfigPool.
+func (ctrl *Controller) finalizeOptOut(mosc *mcfgv1alpha1.MachineOSConfig, mosb *mcfgv1alpha1.MachineOSBuild, pool *mcfgv1.MachineConfigPool) error {
+	if err := ctrl.postBuildCleanup(mosc, mosb, true); err != nil {
+		return err
 	}
+	return nil
+}
 
-	// Perform the post-build cleanup.
-	if err := ctrl.postBuildCleanup(pool, false); err != nil {
-		return fmt.Errorf("could not do post-build cleanup: %w", err)
-	}
+// Marks a given MachineConfigPool as build successful and cleans up after itself.
+func (ctrl *Controller) markBuildSucceeded(mosc *mcfgv1alpha1.MachineOSConfig, mosb *mcfgv1alpha1.MachineOSBuild) error {
+	klog.V(4).Infof("Build %s succeeded for MachineConfigPool %s, config %s", mosb.Name, mosc.Spec.MachineConfigPool.Name, mosb.Spec.DesiredConfig.Name)
 
-	// Perform the MachineConfigPool update.
 	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
-		mcp, err := ctrl.mcfgclient.MachineconfigurationV1().MachineConfigPools().Get(context.TODO(), ps.Name(), metav1.GetOptions{})
+		// REPLACE FINAL PULLSPEC WITH SHA HERE USING ctrl.imagebuilder.FinalPullspec
+		ibr := newImageBuildRequest(mosc, mosb)
+		digestConfigMap, err := ctrl.kubeclient.CoreV1().ConfigMaps(ctrlcommon.MCONamespace).Get(context.TODO(), ibr.getDigestConfigMapName(), metav1.GetOptions{})
 		if err != nil {
 			return err
 		}
 
-		ps := newPoolState(mcp)
+		sha, _ := ParseImagePullspec(mosc.Status.CurrentImagePullspec, digestConfigMap.Data["digest"])
```

Comment:
issue (blocking): In my work on https://github.com/openshift/machine-config-operator/pull/4333, this error not getting checked made finding a bug here much more difficult.

suggestion: Check the error returned by `ParseImagePullspec()` and do something like this:
```golang
sha, err := ParseImagePullspec(mosc.Status.CurrentImagePullspec, digestConfigMap.Data["digest"])
if err != nil {
	return fmt.Errorf("could not create digested image pullspec from the pullspec %q and the digest %q: %w", mosc.Status.CurrentImagePullspec, digestConfigMap.Data["digest"], err)
}
```

The reason for `%q` here is that if the pullspec or digest is empty, the error message will show empty quotes instead of nothing.

issue (blocking): The aforementioned bug was that `mosc.Status.CurrentImagePullspec` was not set yet. Instead, getting it from the MachineOSBuild allows this to happen sooner and not encounter an error along the way. Changing `ParseImagePullspec(mosc.Status.CurrentImagePullspec, digestConfigMap.Data["digest"])` to `ParseImagePullspec(mosb.Status.CurrentImagePullspec, digestConfigMap.Data["digest"])` will alleviate this.

Reply:
This is a very very valid concern, Zack! And is probably the cause of the blocking issue we re facing rn. The source should be either mosb.Status.CurrentImagePullspec or mosc.Spec.BuildInputs.RenderedImagePushspec but not the one in the code (egg-chicken-egg problem) Tysm!